### ISSUE TITLE
emClient 8 blocked by Tine 2.0

### ISSUE THREAD

**Issue by mrHJS** [tine20/tine20#7202]
Hi there,

the new version 8 of emClient is blocked by Tine. When I try to add or edit appointments I get a "forbidden" error. We had this problem a few years ago when version 7 of emClient was published. If I remember correctly, the solution was to add a new emClient7.php file... is it possible to do exactly that for the new version as well?

Thank you in advance!

**Comment by lab-at-nohl** [comment]
Hi, 

emClient 8 looks promising, unfortunately windows only (Ii may start a VirtualBox...). Indeed, this issue is related to #5842. However, we need to identify emClient by its **user agent** string. Do you know it? Can you look into the Tine 2.0 log file (level DEBUG) to find out? Is it like `emClient/8.0.xxxx.x`?

I am happy to help. We either have to find a solution for version 8 or - if only the user agent changed - make that known to Tine 2.0.

**Comment by mrHJS** [comment]
Hi,
thanks for your answer. I looked it up, the user agent is: eMClient/8.0.3318.0
But can't there be like an abstract solution so that it only checks for "eMClient" as this remains the same? It's very annoying having to change that string manually for each new version.

**Comment by lab-at-nohl** [comment]
Yes that is annoying, although probably not entirely Tine 2.0's fault. Did you copy&paste the string (can you make sure)? Please note the little differences in lower/upper case and spaces... As far as I can see in the files, originally (version 5) it was written `eM Client`, then the Beta of version 7 had the string `MailClient`, which was reverted to `eM Client` in the end (probably final version 7 was handled by the original class therefore). 

Now, it seems to be `eMClient` (w/o a space)? Can you double check that no space exists between eM and Client! Also, have you noticed any limitations or mixture of fields with version 7 (hence, can Tine 2.0's class for version 5 also handle version 7 and 8)?

**Comment by mrHJS** [comment]
Yes, I copy/pasted the String from the Logfile...

**Comment by lab-at-nohl** [comment]
Ok. Can you help me test a simple solution?

Please change in three files the line corresponding to the following:
- Calendar/Convert/Event/VCalendar/EMClient.php (line 21)
- Tasks/Convert/Task/VCalendar/EMClient.php (line 23)
- Addressbook/Convert/Contact/VCard/EMClient.php (line 21)

```
    const HEADER_MATCH = '/e[Mm] ?Client\/(?P<version>.*)/';
```

This is a general approach not depending on the version of eM Client that will recognize all user agents used. If it will work as expect,ed I will provide a patch and pull request to the Tine 2.0 devels. 

Then you have to test: Create in eM CLient a contact, an event and a task and fill all fields. Compare it in Tine 2.0 web interface. 

All the Best
Johannes

**Comment by pschuele** [comment]
thanks for opening an issue @mrHJS  and providing a fix @lab-at-nohl :)
I'll review and merge the PR, when the solution is tested.

**Comment by pschuele** [comment]
@mrHJS did you find the time to test the patch by @lab-at-nohl ?

**Comment by mrHJS** [comment]
First of all - I am sorry for my long absence due to all that Corona chaos.
I haven't tested it yet because after finally upgrading to BigSur a few weeks ago, I am still waiting for the announced emClient 8.1 version for MacOS - it's annoying that they have only published the Windows version. But in order to not have to wait any longer, I will test it against the latest 8th version and will come back tomorrow with my results...

**Comment by mrHJS** [comment]
OK, I have already tested it and with the aforementioned line

` const HEADER_MATCH = '/e[Mm] ?Client\/(?P<version>.*)/';`

everything (Contacts, Calendars, Tasks) works as expected - at least with V 8.0.3494, which is the latest macOS version. Thank you very much for your work and effort!!!

**Comment by pschuele** [comment]
> everything (Contacts, Calendars, Tasks) works as expected - at least with V 8.0.3494, which is the latest macOS version. Thank you very much for your work and effort!!!

great, thanks for testing! just merged the PR by @lab-at-nohl

**Comment by sgraf812** [comment]
https://www.tine20.com/fileadmin/docs_tine/tine20_ReleaseNotesFlyer_2020_DE.pdf says there is CalDAV support for emClient 8.0. Actually that doesn't seem to be the case, does it?

**Comment by lab-at-nohl** [comment]
> CalDAV support for emClient 8.0. Actually that doesn't seem to be the case, does it?

How comes, have you tried?

Actually the aforementioned patch added support for emClient 8.0, I thought. However, I could only test with the demo version of emClient back then (a thirty days trial).

**Comment by sgraf812** [comment]
Yes, I tried. TBF, I think our server runs the business version. 8.2 doesn't work, 8.0 doesn't work, but 7.2 works, so I'm running that now. I figured I should post here in the related issue, seeing that it has been closed in Feb while the release notes from last Nov claim it works with emClient 8.0.

**Comment by lab-at-nohl** [comment]
It would have been better to open a new issue since this one is closed ;-)

All versions of emClient should work. And if not you should at least have access read only. 

Can you help and tell aus the **user agent** string of emClient 8.2? Do you know how to write a debug log?

@pschuele Can you reopen the issue?

**Comment by sgraf812** [comment]
To be more explicit than in my last comment: I think it's probably fixed on master, but the linked release notes https://www.tine20.com/fileadmin/docs_tine/tine20_ReleaseNotesFlyer_2020_DE.pdf are for a version from Nov 2020 in which the fix couldn't possibly have landed, because it was fixed in Feb 2021. So it's a matter of backporting the change or amending the release notes to say that 8.0 is not supported.

**Comment by pschuele** [comment]
re-opened the issue.

@sgraf812 could you please provide us the user agent of your emClient?

**Comment by sgraf812** [comment]
> @sgraf812 (or someone else) could you please provide us the user agent of your emClient?

Not sure actually, but according to this answer: https://github.com/tine20/tine20/issues/7202#issuecomment-675577943 the user agent seems to be `eMClient/8.0.3318.0`, with a capital M. I'm using eM Client 7.2 for the time being, which works with the business version. Note that eM Client < 8 appear to use a user agent string starting with `emClient`, e.g., lowercase m. And as I said, the issue probably has been fixed for the community edition, but not for the business edition that we are using. Hence the release notes for the business version are simply inaccurate: It only supports CalDAV write access from eM Client < 8 at the moment. At least that is the behavior I'm observing.

To be clear, a fix would be to either

- Amend the release notes for https://www.tine20.com/fileadmin/docs_tine/tine20_ReleaseNotesFlyer_2020_DE.pdf, or
- Backport https://github.com/tine20/tine20/commit/dc569cc7305848aef2ce1890f74f95e3bacfb143 to the business edition

I also wonder why user agents are whitelisted to begin with. Why can't it just fall back to a generic backend with write access if no user agent matches? That way, I could also use Outlook or Geary (say).

**Comment by pschuele** [comment]
hi,
sorry for the late reply.
as the current business edition (2021.11) supports emclient 8, I think we can close this issue again.

> I also wonder why user agents are whitelisted to begin with. Why can't it just fall back to a generic backend with write access if no user agent matches? 

we did this to prevent data corruption - in the early days, Xdav clients sent broken requests that broke events/contacts in tine... so we decided to have a more defensive approach here.

but maybe now it is the time to introduce a feature switch which allows admins to activate write access for any clients? but this should be put into another issue/feature request.
